Dismemberment implementation

Badulundabad opened the pull request.

hunharibo reviewed (approved):
Looks good.

tristanka reviewed (approved):
Looks good, approved

hunharibo committed:
Merge pull request #1 from WarhammerTheOldWorld/development

bumped to 1.6.0 version in submodule.xml

hunharibo committed:
Merge pull request #6 from WarhammerTheOldWorld/development

Development

Badulundabad committed:
Initial commit

Added decapitation;
There is a problem when setting the invisibility of the chopped off head it affects the heads of the rest of the agents who use this mesh.

Badulundabad committed:
Fix the GetHeadCopy method

Now this method returns the head with helmet or hat;
Fixed a bug when head meshes of all agents disappeared when getting a head mesh of a victim.

hunharibo commented on CSharpSourceCode/HarmonyPatches/MeleeHitCallbackPatch.cs:
```diff
@@ -0,0 +1,20 @@
+            Dismemberment.AttackCollision = collisionData;
+            bool isPDV = collisionData.VictimHitBodyPart == BoneBodyPartType.Head && collisionData.StrikeType == 0 && collisionData.DamageType == 0 && (attacker.AttackDirection == Agent.UsageDirection.AttackLeft || attacker.AttackDirection == Agent.UsageDirection.AttackRight);
+            if (isPDV)
+                Dismemberment.AddPDV(victim, attacker);
```
This should not be a harmonypatch. You should be using a MissionLogic and hook into the OnScoreHit() override method.

hunharibo commented on CSharpSourceCode/SubModule.cs:
```diff
@@ -68,6 +68,82 @@ protected override void OnSubModuleLoad()
             LoadFontAssets();
+            LoadDismembermentSettings();
+        }
+        protected override void OnApplicationTick(float dt)
```
As mentioned above, you should be using a MissionLogic and override its OnMissionTick() method for this. Application Tick is not the place for this. Added benefit if you do it that way is that while in a Mission, you can be sure that Game.Current and Mission.Current exists guaranteed, no need for NPE check.

Badulundabad committed:
Remove Harmony patch

MeleeHitCallbackPatch and PotentialVictim classes were removed;
DismembermentMissionLogic class was created, all logic of condition checking now here.

hunharibo reviewed (approved):
Changes to structure look good.

tristanka commented on CSharpSourceCode/Battle/Dismemberment/Dismemberment.cs:
```diff
@@ -0,0 +1,158 @@
+            MakeHeadInvisible(victim);
+            GameEntity head = SpawnHead(victim);
+            AddPhysicsFromCollision(head, attackCollision);
+            CreateBloodBurst(victim);
```
Move this method and its dependent methods into the `MissionLogic`? I can't imagine this should be used outside of the dismemberment mission logic, especially since this should only work within missions. Also, if we don't need the commented out code on lines 18 and 19, I think we should just go ahead and remove them

Z3rca replied:
agree that could be all inside the same mission logic. I dont understand the difference here. I would move the class if the Inheritage of another tale worlds class , is needed otherwise , if the function are that related, keep them together

tristanka commented on CSharpSourceCode/Battle/Dismemberment/Dismemberment.cs:
```diff
@@ -0,0 +1,158 @@
+            MatrixFrame boneEntitialFrameWithIndex = victim.AgentVisuals.GetSkeleton().GetBoneEntitialFrameWithIndex((byte)victim.BoneMappingArray[HumanBone.Head]);
+            Vec3 vec = victim.AgentVisuals.GetGlobalFrame().TransformToParent(boneEntitialFrameWithIndex.origin);
+            victim.CreateBloodBurstAtLimb(13, ref vec, 0.5f + MBRandom.RandomFloat * 0.5f);
+        }
```
I think we should move these to `MissionLogic` and remove the `static` modifier (could just remove that for all of the methods within the mission logic)

tristanka commented on CSharpSourceCode/Battle/Dismemberment/Dismemberment.cs:
```diff
@@ -0,0 +1,158 @@
+            agentBuildData.InitialFrame(frame);
+
+            return Mission.Current.SpawnAgent(agentBuildData, false, 0);
+        }
```
This could be moved into a Utilities class called something like `TowDebug` if we want to keep it. I don't think it fits to be in a dismemberment-specific class but could be useful in the future.

Z3rca replied:
agree use other Debug Functions , instead of creating those.

tristanka commented on CSharpSourceCode/Battle/Dismemberment/Dismemberment.cs:
```diff
@@ -0,0 +1,158 @@
+                blow.DamageCalculated = true;
+                agent.RegisterBlow(blow);
+            }
+        }
```
This method looks like it's unused. If it is, I think we should go ahead and remove it.

Z3rca replied:
agree

tristanka commented on CSharpSourceCode/Battle/Dismemberment/Dismemberment.cs:
```diff
@@ -0,0 +1,158 @@
+        private static void DisplayMessage(string msg, uint color)
+        {
+            InformationManager.DisplayMessage(new InformationMessage(msg, Color.FromUint(color)));
+        }
```
These methods are basically handled in `TOWCommon` - I think they should be removed here. If we want to extend debug message support, we should add functionality to `TOWCommon`. (Also, with the mention of `TOWDebug` earlier, maybe `TOWCommon.Say()` should be moved to the debug class anyway)

Z3rca replied:
as explained above, use the TowCommon class and make debug messages. anyway this information shouldn't be shown anyway, and should be deactivated by default, I would say.

tristanka commented on CSharpSourceCode/Battle/Dismemberment/DismembermentMissionLogic.cs:
```diff
@@ -0,0 +1,39 @@
+        {
+            base.OnRegisterBlow(attacker, victim, realHitEntity, blow, ref collisionData, attackerWeapon);
+
+            bool isCompliant = victim.IsHuman &&
```
This may be more clearly named as `bool shouldDismember` or something similar. `isCompliant` is a little ambiguous

Z3rca replied:
agree

tristanka commented on Settings.cfg:
```diff
@@ -0,0 +1,5 @@
+// Settings configuration file
+Chance=1
+SlowMotion=false
+Debug=true
```
I don't see that this file is being loaded anymore. Can we delete this Settings.cfg or should we be loading it in still? (or am I missing where it's loaded?)

Z3rca replied:
reminds me of endourse jedijosh920. We have to state, that our improvement is based on his original mod.

Z3rca replied:
did this in the readme pull request

Z3rca reviewed (changes requested):
check out the changes I made. It is basically the aproval of @tristanka. Do this changes and we are good to merge I would say. the Interconnection with other systems are not that problematic I would say, so we can extent this mod. 

Also I will mention jedijosh920 in the readme as endoursement . - Please make sure, that it is stated clearly, where you copied, and where you extended, maybe in the description of the pull request ( also maybe in some kind of method description on the head) , remove unnessary code which are not needed for us anymore but was in the original mod. We should maybe, also mention or release this dismemerment as sub mod, with offcourse the mentioning of jedijosh as our reference.

Badulundabad committed:
Added a lot of changes. Read the description.

Added the meshes are covering the place where the head is cut off;
Now a hat falls separately if an agent has it;
The physics of the fall of the hat and head has been adjusted;
The dismemberment chance formula has been adgjusted.

Badulundabad committed:
Added a lot of changes. Read the description.

The DismembermentMissionLogic and Dismemberment classes have been united;
The static modifiers have been removed from the all methods;
The SpawnAgent method has been moved to the TOWDebug class;
The KillAgent method has been removed;
The calls of the DisplayDebugMessage method have been replaced by TOWCommon.Say();
The DisplayDebugMessage has been removed;
The isCompliant variable has been renamed to canBeDismembered;
The Settings.cfg file has been removed.

Badulundabad committed:
Removed the debug fields

Badulundabad committed:
Added variables for mod customization
